### PR TITLE
laters.recureent: fix masking from input

### PR DESCRIPTION
**Pull request by amitbeka**
If there is no input layer to a recurrent layer, mask is automatically
taken from the input data. Fix a bug where timesteps whose features were
all zeroes were included in the mask, instead of excluded.

Signed-off-by: Amit Beka <amit.beka@gmail.com>

### CODE REVIEW

**Comment by amitbeka** [comment]
Solves issue #355.

**Comment by fchollet** [comment]
Can you sum up the context of this: what was the issue, what were some possible options to fix it, and what you decided to do? I'm out of the loop here.

**Comment by wxs** [comment]
Hmm. This is treating all-zero inputs to recurrent layers as masked, which is incorrect in the case where you are NOT using masking. For instance if I had am `Embedding -> SimpleRNN ->SimpleRNN`, and my  Embedding was not masking, but my first SimpleRNN used ReLUs, it could easily happen to send an all-zero input to the second SimpleRNN which would then "turn it" into a mask.

If you want to introduce a mask, I think you would be better to add a flag to the recurrent layers similar to the "mask_zero" flag of Embedding that is False by default. Then in the event that your first layer was an RNN layer and you wanted it to introduce masks, you would just set the mask on that layer.

**Comment by amitbeka** [comment]
OK, sounds good enough. @melonista is checking the current implementation now, and I will look at it too. maybe this patch isn't necessary, but if so, I'll add the `mask_zero` flag

**Comment by fchollet** [comment]
I took a look at this, correct me if I'm getting it wrong, but I think it's nonsense.

**Comment by amitbeka** [comment]
I think you do getting it wrong - there is no way to have padding in recurrent layers right now, except if the previous layer was Embeddings. `fit` expects 3-dimentional matrix, so it must be padded somehow, and there is no way to indicate what is the padding.

**Comment by wxs** [comment]
@fchollet I think this PR makes sense, actually, for people who have temporal continuous data (vs. temporal discrete data embedded via Embedding which was my use case)

**Comment by wxs** [comment]
Although maybe it would be more clear to have an explicit `Mask` layer, that introduces the mask, rather than having it internal to the recurrent layers.

**Comment by wxs** [comment]
Though some of the same concerns we had about what value to use as the mask signal apply, and perhaps there's a completely different approach that solves that.

**Comment by fchollet** [comment]
Ok, I see. I think we do need a way to pass masks as input to a model. Maybe a special Layer looking for certain sentinel values? (not very elegant, though).

**Comment by amitbeka** [comment]
For temporal masking (padding timesteps), I think the most simple and effective solutions are either this one (use `mask_zero`), or having a `Padding` layer, which accepts list of 2-d matrices and list of arrays/matrices as `X` and `y`, and outputs two matrices which are padded, along with setting `self.output_mask` for the next layer to read.

The more I think of it, the `Padding` layer idea seems better (instead of mixing the padding-handling into `Recurrent`, we separate it to a different layer).

Using sentinel values in a `Masking` layer (which already accepts the data padded, and generates a mask) is a little more general, but I think for 99% of the cases it will create more work for the user than this option.

Let's decide on this so I can go on and implement.

**Comment by amitbeka** [comment]
Related issue: I think that TimeDistributedDense doesn't apply the mask it receives from its input layer at any stage? am I right? not sure what is the elegant fix though..

**Comment by wxs** [comment]
> Related issue: I think that TimeDistributedDense doesn't apply the mask it receives from its input layer at any stage? am I right? not sure what is the elegant fix though..

But this doesn't matter for TimeDistributedDense, no? Because it doesn't share information between timesteps, and it also passes the mask through unchanged for the next layer. So it produces "garbage" values at masked timesteps, but there is no "correct" values for it to produce there anyway.

**Comment by wxs** [comment]
> or having a Padding layer, which accepts list of 2-d matrices and list of arrays/matrices as X and y, and outputs two matrices which are padded, along with setting self.output_mask for the next layer to read.

That seems problematic because it changes the interface to `fit()`, no? Instead of it taking `(X, y)` it needs to take like `(X, y, pad)` or something.

**Comment by amitbeka** [comment]
@wxs you are correct, it will make a change to `fit()` which we probably don't want. Instead I've decided to implement a `Masking` layer which gets a sentinal value and sets the mask to zero wherever it is found (like you've suggested in previous PR #244 ).

From what I have understood there, I need to use `sample_weight` in `fit()` so the loss function will return zero for padded values, but I can't seem to make things work - it makes results worse. Moreover, using non-zero values for padding either the samples (features) or the targets make the results much worse, which I think means that the samples/targets aren't really masked when training.

I have pushed the test script (which is documented for you to play with) and the full results i've got as a new commit here, but here is a summary of my findings:

padding | use Masking layer | use mask_zero | use sample_weight | x_mask_value | y_mask_value | accuracy
----- | ----- | ----- | ---- | ----- | ---- | -----
none | no | no | no | | | 0.9947 (this is the baseline)
pre | no | no | no | 0 | 0,0 | 0.7922 (fails, expectedly)
post | no | no | no | 0 | 0,0 | 0.847 (also expected)
pre | yes | no | no | 0 | 0,0 | 0.9939 (works, but without weighting - strange)
post | yes | no | no | 0 | 0,0 | 0.7712 (doesn't work with post-padding and without weights)
pre | yes | no | yes | 0 | 0,0 | 0.7448 ( doesn't work with weighting, althoug it should)
pre | yes | no | yes | 0 | 1,1 | 0.9955 (but works with different target padding!)
post | yes | no | yes | 0 | 0,0 | 0.9552 (what's happening?)
post | yes | no | yes | 100 | 0,0 | 0.7481 (fails with non-zero features)
pre | no | yes | yes | 0 | 0,0 | 0.9980 (works correctly)
pre | no | yes | yes | 0 | 1,1 | 0.9960 (still correct)
pre | no | yes | no | 0 | 1,1 | 0.9948 (still fairly correct, without weighting)
post | no | yes | yes | 0 | 0,0 | 0.9743 (moving to post-padding hurt results)
post | no | yes | yes | 0 | 1,1 | 0.9723 (same with different targets)

**Comment by amitbeka** [comment]
Because the PR is closed it doesn't show my script and updates, they can be pulled from:
https://github.com/amitbeka/keras/tree/fix-recurrent-masking

**Comment by wxs** [comment]
OK well skimming the code on your branch, I see some issues. Your `Masking` return something of the same shape as X, but it should return something of with one less dimension, i.e. (nb_samples, nb_timesteps, nb_dims) -> (nb_samples, nb_timesteps)

I can't look at it more closely right now, but this is a signal that something might be wrong.

**Comment by amitbeka** [comment]
My masking does the same as `Embedding` does - it returns the same shape as the data, and `Recurrent.get_padded_shuffled_mask` does the dimensionality reduction, broadcasting, and dimshuffling. I did try to return a mask of (nb_samples, nb_timesteps), and it seems to return the same results (see the branch for the change, results are in file padding2.log).

narrowing on the problem:
- using `Masking` layer with `sample_weight`, pre-padding, y values as [0,0] doesn't work, although I think it should. Changing to post padding, [1,1] or remove `sample_weight` "solves" it.

- using `Masking` layer with `sample_weight`, post-padding, and y value of [0,0] works when x padding value is 0, but not when it's 100

- using `mask_zero` inside the recurrent layer works fine with pre-padding, but have slightly worse results with post padding

**Comment by amitbeka** [comment]
@wxs and @fchollet -- I wonder how we can progress with this patch; I use the patched code successfully with my project (and as stated before, this is the only option for padding without an Embedding layer), and would be glad to contribute
